Add support for libvips in addition to ImageMagick

Gargron opened the pull request:
Libvips can be used instead of ImageMagick with `MASTODON_USE_LIBVIPS=true`. Enabled by default in the Docker image.

- [x] Removes metadata
- [x] Retains color profile
- [x] Reads pixels for blurhash
- [x] Reads histograms for color extraction
- [x] Crops animated GIFs when needed

___

Fixes MAS-223

Gargron committed:
Change image processing from ImageMagick to libvips

shleeable commented on lib/paperclip/lazy_thumbnail.rb:
```diff
@@ -1,20 +1,50 @@
 
-      Paperclip::Thumbnail.make(file, options, attachment)
+      vips.convert(@format)
+          .saver(interlace: true, quality: 90)
```
Interlaced defaults to false. Is this recommended?

renchap replied:
Yes, this will make Vips output progressive JPEGs, which are slightly (a few percent) larger than normal JPEGs, but will start displaying data much quicker.

shleeable replied:
Love that.. Shame something newer formats seem to be missing progressive atm. progressive AVIF seems to be on the way.

renchap reviewed (commented):
The Github actions will need to install `libvips42` instead of `imagemagick`.

There are probably some ImageMagick related files to remove, for example `config/imagemagick`.

The Docker image will also need to be changed to install vips, @vmstan started to work on it. Maybe you can update it in this PR to simply install `libvips42` in place of `imagemagick`, and @vmstan can open another PR later to get it installed better (because the current Debian package enabled far too many formats for our usage).

We will also need to take great care in the upgrade instructions to tell people that they need to install VIPS now (and uninstall IM?). This will also be an issue with people running on nightly. The alternative is to support both IM and VIPS for 4.3, with a config option to enable VIPS, show a deprecation warning with IM is used, and remove IM support in 4.4, but I do not know if this is worth it.

renchap commented:
It is also probably a good idea to call `Vips.block_untrusted` (see https://github.com/libvips/ruby-vips/pull/382) so only the trusted loaders (ie, that are covered by the fuzzing testing) are enabled. 

Untrusted operations can be seen with `vips -l | grep untrusted`. There are more details in the [8.13 release notes](https://www.libvips.org/2022/05/28/What's-new-in-8.13.html).

renchap commented on .github/workflows/check-i18n.yml:
```diff
@@ -21,6 +21,8 @@ jobs:
 
       - name: Set up Ruby environment
         uses: ./.github/actions/setup-ruby
+        with:
```
If `libvips42` needs to be installed for every workflow, then it is best to add it to `.github/actions/setup-ruby/action.yml` and not use `additional-system-dependencies`.

The alternative is only requiring vips on the Ruby side in files where it is used, but I think I prefer everything to fail if vips is not installed, so admins will know if their setup is wrong.

shleeable commented:
Worth adding libheif?

vmstan commented:
> Worth adding libheif?

It's included by the libvips42 package.

vmstan commented:
I’ve been running this PR for the last 20 hours on [vmst.io](http://vmst.io/) and it’s been largely successful. I have had no user reported issues uploading and processing image content, but noticed a few entries in Sidekiq dead queue triggered by a few remote posts.

### `extract_area: bad extract area`
https://plasmatrap.com/notes/9t1auqkbl0
(There are many entries all related to this note via replies, etc, but this is the source post)

Related backtrace:
```
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:228:in `build'
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:481:in `call'
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/image.rb:229:in `method_missing'
```

### `hist_find_ndim: image is not 1 - 3 bands`
https://graphics.social/users/metin/statuses/112404645439099091
https://mastodon.online/users/spocko/statuses/112401772350190886
https://mastodon.social/users/wearenew_public/statuses/112401409861904050

Related backtrace:
```
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:228:in `build'
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:481:in `call'
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/image.rb:229:in `method_missing'
```

### `VipsForeignSave: "/tmp/c0ddd4cf006863154a6ba8d7b30396bd20240507-7-ivyc3w.jfif" is not a known file format`
https://mastodon.art/users/catrionaroberts/statuses/112400639207525684

Related backtrace:
```
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/image.rb:597:in `write_to_file'
/opt/mastodon/lib/paperclip/lazy_thumbnail.rb:47:in `make'
/usr/local/bundle/gems/kt-paperclip-7.2.2/lib/paperclip/processor.rb:33:in `make'
```

The JFIF file format error appears to be related to https://github.com/libvips/libvips/issues/3775 which was fixed by a later version of libvips than what comes from the Debian repo. After this error appeared I switched to using the latest version of libvips compiled into the Docker container using https://github.com/vmstan/mastodon/pull/27 which will be submitted as a change after this PR has merged.

vmstan commented:
As far as performance data is concerned, so far I would say libvips has led to slightly lower container CPU utilization. I don't have any fine grained metrics on image processing performance, but comparing the last day worth of traffic to the same time period in the previous week, which had similar amounts of Sidekiq traffic processed:

Tuesday 7 into Wednesday 8 May 2024
![CleanShot 2024-05-08 at 09 04 34@2x](https://github.com/mastodon/mastodon/assets/3002053/65d3805f-a6b0-418b-b997-2a8b1731016e)

Tuesday 30 April into Wednesday 1 May 2024
![CleanShot 2024-05-08 at 09 05 49@2x](https://github.com/mastodon/mastodon/assets/3002053/88251b10-d39b-4858-9f9b-49bfce6b4c33)

Average CPU usage (blue line) appears consistently lower, as do all load averages. 

These are on three, 4x dedicated vCPU droplets at Digital Ocean. Average Sidekiq jobs per day approx 1.4 million.

vmstan commented:
I've run into an issue where uploading a PNG file that exceeds the allowed dimensions/megapixels results in an error of `422 Validation failed: File must be less than 16 MB, File file size must be less than 16 MB` even if the file itself is less than 16 MB.

![CleanShot 2024-05-09 at 16 09 20@2x](https://github.com/mastodon/mastodon/assets/3002053/84b6aaaa-1a0e-48a8-bcc8-ac9005bd86a1)

Uploading a file that is exactly 2160x3840 works, but a file that is 2161x3842 will result in the above error. In both cases the file is roughly 4.6 MB in size. So not only is the wrong type of error being triggered, it doesn't seem like any error should be triggered if the file was rescaled by vips.

I do not see this issue with JPG files.

LeoEurope commented:
> I’ve been running this PR for the last 20 hours on [vmst.io](http://vmst.io/) and it’s been largely successful. 

@vmstan As a switch to libvips supposedly allows WebP and AVIF files to be rendered in link previews (closing https://github.com/mastodon/mastodon/issues/27370 and https://github.com/mastodon/mastodon/issues/14983, and probably many more) do you see that many of the issues listed [here](https://github.com/mastodon/mastodon/issues/14983#issuecomment-1520594700) are resolved?

vmstan commented:
> > I’ve been running this PR for the last 20 hours on [vmst.io](http://vmst.io/) and it’s been largely successful.
> 
> @vmstan As a switch to libvips supposedly allows WebP and AVIF files to be rendered in link previews (closing #27370 and #14983, and probably many more) do you see that many of the issues listed [here](https://github.com/mastodon/mastodon/issues/14983#issuecomment-1520594700) are resolved?

I test posted all of the links shared in that comment and they looked the same on my instance running with vips as they do on mastodon.online running ImageMagick and similar build of 4.3.

LeoEurope commented:
> > > I’ve been running this PR for the last 20 hours on [vmst.io](http://vmst.io/) and it’s been largely successful.
> > 
> > 
> > @vmstan As a switch to libvips supposedly allows WebP and AVIF files to be rendered in link previews (closing #27370 and #14983, and probably many more) do you see that many of the issues listed [here](https://github.com/mastodon/mastodon/issues/14983#issuecomment-1520594700) are resolved?
> 
> I test posted all of the links shared in that comment and they looked the same on my instance running with vips as they do on mastodon.online running ImageMagick and similar build of 4.3.

OK, that's odd. As I've already bug spammed this ticket enough and this is a libvips implementation ticket let's get to the bottom of the issue using ticket https://github.com/mastodon/mastodon/issues/14983 and use ticket https://github.com/mastodon/mastodon/issues/27370 as a potential stopgap until the system accepts WebP and AVIF. Once libvips has properly landed I can be more helpful with troubleshooting.

vmstan commented:
CPU usage before and after over a 14 day period.
![CleanShot 2024-05-13 at 09 11 12@2x](https://github.com/mastodon/mastodon/assets/3002053/ff83303b-cd36-4efe-90cf-675de9f1e55e)

renchap commented:
This seems to be a significant improvement overall (and less spikes?), really promising!

vmstan commented:
I have flipped back to libvips since the rebase yesterday, and so far have not see seen any recurrence of the three errors in https://github.com/mastodon/mastodon/pull/30090#issuecomment-2100558303 but the PNG uploads continue to be an issue. Also got this one today:

`Vips:Error: extract_band: bad extract band`
https://mastodon.social/users/pkashwin/statuses/112477913618386753

```
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:228:in `build'
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:481:in `call'
/usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/image.rb:229:in `method_missing'
```

renchap commented:
> I have flipped back to libvips since the rebase yesterday, and so far have not see seen any recurrence of the three errors in [#30090 (comment)](https://github.com/mastodon/mastodon/pull/30090#issuecomment-2100558303) but the PNG uploads continue to be an issue. Also got this one today:
> 
> ### `Vips:Error: extract_band: bad extract band`
> https://mastodon.social/users/pkashwin/statuses/112477913618386753
> 
> ```
> /usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:228:in `build'
> /usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/operation.rb:481:in `call'
> /usr/local/bundle/gems/ruby-vips-2.2.1/lib/vips/image.rb:229:in `method_missing'
> ```

Can you share your PNG test file (here or privately on Discord)?

vmstan commented:
Sure, it's a just an image I grabbed off Unsplash and then converted to PNG using Pixelmator. The one ending in 83 is the exact 8.3 megapixel limit and will upload, the one ending in 84 generates the error.


![aditya-chinchure-4z0Lz0Fn6oU-unsplash_83](https://github.com/mastodon/mastodon/assets/3002053/5814345c-43a3-4667-b5cf-f9e0699d8fdb)
https://github.com/mastodon/mastodon/assets/3002053/5814345c-43a3-4667-b5cf-f9e0699d8fdb

![aditya-chinchure-4z0Lz0Fn6oU-unsplash_84](https://github.com/mastodon/mastodon/assets/3002053/0b528111-e911-4af1-ab9f-9e300e0a13fc)
https://github.com/mastodon/mastodon/assets/3002053/0b528111-e911-4af1-ab9f-9e300e0a13fc

vmstan commented:
Another instance of `Vips:Error: extract_band: bad extract band`
https://post.lurk.org/users/sean_ae/statuses/112485874698250947

ClearlyClaire reviewed (changes requested):
I still need to review `LazyThumbnailer` and re-read `ColorExtractor#palette_from_image`, but this looks good overall.

I think I found a bug in `BlurhashTranscoder`, left an inline comment.

ClearlyClaire commented on lib/paperclip/color_extractor.rb:
```diff
@@ -78,6 +86,38 @@ def make
+      end
+    end
+
+    def palette_from_image(image)
+      histogram = image.hist_find_ndim(bins: BINS)
+      _, colors = histogram.max(size: 10, out_array: true, x_array: true, y_array: true)
+
+      colors['out_array'].map.with_index do |v, i|
+        x = colors['x_array'][i]
+        y = colors['y_array'][i]
+
+        rgb_from_xyv(histogram, x, y, v)
+      end.reverse
+    end
+
+    # rubocop:disable Naming/MethodParameterName
+    def rgb_from_xyv(image, x, y, v)
+      pixel = image.getpoint(x, y)
+      z = pixel.find_index(v)
+      r = (x + 0.5) * 256 / BINS
+      g = (y + 0.5) * 256 / BINS
+      b = (z + 0.5) * 256 / BINS
+      ColorDiff::Color::RGB.new(r, g, b)
+    end
```
If my understanding of libvips histograms is correct, I think `rgb_from_xyv` is unsound if there are two `b₁` and `b₂` values such that `(r, g, b₁)` and `(r, g, b₂)` appear in the same frequency. If that is the case, `rgb_from_xyv` will not be able to distinguish `b₁` and `b₂`. I can't think of an easy way to fix this, though.

I'm not completely sure this is the most appropriate tool for the task.

Gargron replied:
The code is adapted from an example provided by libvips' author: https://gist.github.com/jcupitt/ee3afcbb931b41b4d7f4

ClearlyClaire replied:
Sure, but I think my point still stands. I'm not sure how to fix that, though.

ClearlyClaire replied:
I think that code is indeed incorrect, but the failing cases are probably very unlikely (multiple of the ten top color bins only differing on the blue color component *and* having the same exact occurrence count).

I refactored the code to make it clearer and added comment to explain the non-obvious way histograms are modeled, as well as highlight this bug.

I don't know how to fix it without basically reimplementing `max` ourselves, so I'll leave it at this for the moment.

ClearlyClaire commented on config/initializers/vips.rb:
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+
+Vips.block_untrusted(true) if Vips.at_least_libvips?(8, 13)
```
Maybe instead or in addition to that, it could make sense to only enable loaders we know we will use?

Something like:
```ruby
Vips.block('VipsForeign', true)
%w(VipsForeignLoadJpeg VipsForeignLoadPng VipsForeignLoadWebp VipsForeignLoadHeif VipsForeignSavePng VipsForeignSaveJpeg VipsForeignSaveWebp).each do |operation|
  Vips.block(operation, false)
end
```

ClearlyClaire replied:
Unfortunately, `Vips.block` itself is only available for >= 8.13 as well, but I think we should really require this, given that distributions of `libvips` can include unsafe handlers for lots of different file formats.

ClearlyClaire replied:
Solving this conversation with a hard requirement on Vips 8.13+

Gargron commented on lib/paperclip/color_extractor.rb:
```diff
@@ -7,15 +7,23 @@ class ColorExtractor < Paperclip::Processor
+        bottom = image.crop(0, image.height - block_edge_dim, image.width, block_edge_dim)
+        left = image.crop(0, block_edge_dim, line_edge_dim, image.height - (block_edge_dim * 2))
+        right = image.crop(image.width - line_edge_dim, block_edge_dim, line_edge_dim, image.height - (block_edge_dim * 2))
+        top.join(bottom, :vertical).join(left, :horizontal).join(right, :horizontal)
```
Yes, confirmed visually. There are no blank pixels.

ClearlyClaire replied:
Indeed, this should line up exactly.

vmstan commented:
FYI, can't test this PR at the moment because the patch won't apply for my builds because of changes to `app/models/preview_card.rb` even though it's not showing rebase required in GitHub.

ClearlyClaire commented:
> FYI, can't test this PR at the moment because the patch won't apply for my builds because of changes to `app/models/preview_card.rb` even though it's not showing rebase required in GitHub.

Hm, I don't have any merge conflict when merging here.

Ember-ruby commented:
does this fix mastodon's handling of image formats that offer both lossless and lossy?

(upload a lossless webp currently and it will re-encode it to q90 lossy, which looks significantly worse)

ClearlyClaire commented on lib/paperclip/lazy_thumbnail.rb:
```diff
@@ -1,24 +1,128 @@
+          target_height = (@current_geometry.height * scale).round
+        end
+
+        # The only situation where we use crop on GIFs is cropping them to a square
+        # aspect ratio, so this is the only special case we implement. If cropping
+        # ever becomes necessary for other situations, this will need to be expanded.
+        if @target_geometry&.square?
+          crop_width = [target_width, target_height].min
+          crop_height = [target_width, target_height].min
+        end
```
I don't understand when that would happen. Wouldn't the source already be square and the aspect ratio preserved?

Gargron replied:
The source is not necessarily square at all before the crop. The resize operation maintains original aspect ratio.

ClearlyClaire commented on lib/paperclip/lazy_thumbnail.rb:
```diff
@@ -1,24 +1,128 @@
+          target_width = 'iw'
+          target_height = 'ih'
+        else
+          scale = [@target_geometry.width.to_f / @current_geometry.width, @target_geometry.height.to_f / @target_geometry.height].min
```
Typo? Shouldn't this be:
```suggestion
          scale = [@target_geometry.width.to_f / @current_geometry.width, @target_geometry.height.to_f / @current_geometry.height].min
```

ClearlyClaire commented on lib/paperclip/lazy_thumbnail.rb:
```diff
@@ -1,24 +1,128 @@
+        end
+
+        command = Terrapin::CommandLine.new(Rails.configuration.x.ffmpeg_binary, '-nostdin -i :source -map_metadata -1 -filter_complex :filter -y :destination', logger: Paperclip.logger)
+        command.run({ source: @file.path, filter: "fps=#{GIF_FPS},#{filter},split[a][b];[a]palettegen=max_colors=#{GIF_PALETTE_COLORS}[p];[b][p]paletteuse=dither=bayer", destination: dst.path })
```
Why hardcode a GIF framerate? Shouldn't we use the same exact frames as the source?

Gargron replied:
The idea is to keep the file size down so that more avatars/headers fit under the limit.

renchap replied:
Maybe we can set a max framerate (24ips?)

ClearlyClaire replied:
GIFs can have variable framerate, I think providing a fixed FPS may actually increase the resulting file size if the source file has a lower effective average framerate, but this probably needs some more testing

ClearlyClaire replied:
Changed it to a 60 fps max framerate, and a maximum of 3000 frames. Looking at GIF files known to my server, many GIFs are 30, 33 or 50 fps and would feel significantly degraded with a 12 fps hardcoded framerate.

Some GIFs known to my server are 100 fps, but that is most probably a waste we can avoid.

ClearlyClaire committed:
Make libvips opt-in (#30504)

ClearlyClaire commented on lib/paperclip/vips_lazy_thumbnail.rb:
```diff
@@ -0,0 +1,143 @@
+          target_width = 'iw'
+          target_height = 'ih'
+        else
+          scale = [@target_geometry.width.to_f / @current_geometry.width, @target_geometry.height.to_f / @target_geometry.height].min
```
Typo? Shouldn't this be:
```suggestion
          scale = [@target_geometry.width.to_f / @current_geometry.width, @target_geometry.height.to_f / @current_geometry.height].min
```

ClearlyClaire replied:
I think this was a bug, and I fixed it.

ClearlyClaire commented on lib/paperclip/vips_lazy_thumbnail.rb:
```diff
@@ -0,0 +1,143 @@
+          target_height = (@current_geometry.height * scale).round
+        end
+
+        # The only situation where we use crop on GIFs is cropping them to a square
+        # aspect ratio, so this is the only special case we implement. If cropping
+        # ever becomes necessary for other situations, this will need to be expanded.
+        if @target_geometry&.square?
+          crop_width = [target_width, target_height].min
+          crop_height = [target_width, target_height].min
+        end
```
In which case could the `@target_geometry` be square while the source isn't?

Gargron replied:
Yes, absolutely. Sorry if this is written confusingly, crop width and height are set to the exact same value, what I would call `min_side`.

ClearlyClaire replied:
Oh, I understand, this can happen when the target is an avatar. Rewrote it to make it clearer but the logic is sound.

ClearlyClaire committed:
Fix loading of MASTODON_USE_LIBVIPS from dotenv

ClearlyClaire committed:
Fix handling of monochrome images

ClearlyClaire committed:
Guard against possible segfaults

This condition should never happen, but if it does, an exception is better than a segfault

ClearlyClaire committed:
Add monochrome image test

ClearlyClaire committed:
Clarify square case in scaling code

ClearlyClaire committed:
Fix scaling code

ClearlyClaire commented on lib/paperclip/blurhash_transcoder.rb:
```diff
@@ -5,10 +5,9 @@ class BlurhashTranscoder < Paperclip::Processor
+      image = Vips::Image.thumbnail(@file.path, 100)
 
-      attachment.instance.blurhash = Blurhash.encode(geometry.width, geometry.height, pixels, **(options[:blurhash] || {}))
+      attachment.instance.blurhash = Blurhash.encode(image.width, image.height, image.to_a.flatten, **(options[:blurhash] || {}))
```
There was an issue with monochrome images as well as images with transparency. Fixed it with `image.colourspace(:srgb).extract_band(0, n: 3)`

ClearlyClaire commented on config/imagemagick/policy.xml:
Resolving this conversation, as we're keeping this file and making libvips opt-in.

ClearlyClaire committed:
Tweak GIF processing to permit smoother animations and reject longer files

ClearlyClaire commented:
> does this fix mastodon's handling of image formats that offer both lossless and lossy?
> 
> (upload a lossless webp currently and it will re-encode it to q90 lossy, which looks significantly worse)

I don't think this PR changes that significantly, and we want to avoid large file sizes in general. But if you have an example file, I can have a look at what the quality ends up like.

ClearlyClaire reviewed (dismissed):
I am still not 100% sure about the color extraction code, but if my understanding of it is correct, the worst case scenario is rare and just results in suboptimal color selection, not any functional error.

I will continue investigating this, but this should not block a merge.

ClearlyClaire committed:
Refactor color extraction for readability, and add comments

Ember-ruby commented:
> > does this fix mastodon's handling of image formats that offer both lossless and lossy?
> > (upload a lossless webp currently and it will re-encode it to q90 lossy, which looks significantly worse)
> 
> I don't think this PR changes that significantly, and we want to avoid large file sizes in general. But if you have an example file, I can have a look at what the quality ends up like.

webp's 90 quality generally has significantly worse compression artifacts than others at 90 quality as mastodon uses the same quality for all, this is quite an issue for WEBP
not sure if/how this is different in vips, but doing a basic look-through i don't find what quality settings are parsed to it, and with some basic CLI testing, it seems to exhibit the same issues

(a few test images, encoded with imagemagik, same arguments as [`GLOBAL_CONVERT_OPTIONS`](https://github.com/mastodon/mastodon/blob/521b43393350be16fbd1aa003607480c2442f3eb/app/models/media_attachment.rb#L174), except where noted. all images 3607x2300px, scaled down from a larger photo to meet the pixel limit.) 

*this particular scene is probably one of the worst cases for WEBP compression, but it's not exclusive to this case, and it's not that uncommon*

https://bin.blobfox.coffee/file/e7WGQe/mastodon-quality-test-image-reference.png
6.8MiB

https://bin.blobfox.coffee/file/dGZK0b/mastodon-quality-test-image-lossless.webp
4.8MiB - lossless WEBP, *better* compression than PNG

At the moment mastodon will reencode that lossless WEBP into the image below, so if you want to upload something losslessly you will have to upload as a usually less compression efficient PNG, wasting your time, compute, and storage

https://bin.blobfox.coffee/file/dw81re/mastodon-quality-test-image-default.webp
121KiB - what it will be encoded to, quite bad quality, very noticeable artefacts.
(in testing on my instance it produced an even worse quality image than this at ~100KiB)

https://bin.blobfox.coffee/file/ejNxze/mastodon-quality-test-image.jpg
554KiB - quality pretty okay imo, particularly for half a megabyte at ~8.2 Megapixels

https://bin.blobfox.coffee/file/aOo3Ya/mastodon-quality-test-image-jpeg-like.webp
305KiB - this is using the `-define webp:emulate-jpeg-size=true` option, it's better than default but probably still a bit too low

i would guess that checking which method an image in a codec capable of both lossless *and* lossy modes uses would reduce media storage usage, due to the above, and increasing the quality of lossy WEBP to be more visually on par with JPEG would help with getting people to use the more efficient codec, rather than being appalled at their image being immensely compressed. 


uhh, TLDR: lossless images in lossy codecs should be treated like PNG (WEBP at least stores metadata on whether it is or isn't lossless), and WEBP's quality 90 produces much worse looking images than JPEG quality 90

sorry if this is an excessive reply, lol. /gen

mjankowski commented:
I realize this is a month old and merged already, but curious on the added CI config here -- is the only difference between the newly added VIPS section and the regular spec run -- a) the addition of the env var to opt-in to vips, b) running just the paperclip_processing subset of specs on the vips section?

If so ... I wonder if we could just tack this on to the end of the other runs (in ruby version matrix)? Those specs take ~30s, but we're tacking on another ~1min per ruby version for the setup surrounding them.

ClearlyClaire commented:
Yes, those are about the only changes. I suppose moving them to an additional step in the regular tasks may make sense.

mjankowski commented:
Looked a bit more and realized that the VIPS job also switched the runner to `ubuntu-24.04` (the default GH actions `ubuntu-latest` is still 22.x). I assume this gets us the correct/latest vips packages.

I also thought about this a bit more and it might not make sense to tack on more tasks to the end of what is already the longest task ... even though we are repeating the ruby setup, the vips task woul be done before the main long rspec run. So tacking it would reduce the total CI minutes in a given day (fewer over ruby setup runs), but not shorten the length of the run for a single PR, if that makes sense.

May re-visit this when GH promotes ubuntu 24 to be the `-latest` tag.
